<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Intro to scRNAseq and demultiplexing • demultiplexingbioc23</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Intro to scRNAseq and demultiplexing">
<meta property="og:description" content="demultiplexingbioc23">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demultiplexingbioc23</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a-intro-to-demultiplexing.html">Intro to scRNAseq and demultiplexing</a>
    </li>
    <li>
      <a href="../articles/b-hashing-benchmarking.html">Simulated benchmarking of hashing based algorithms</a>
    </li>
    <li>
      <a href="../articles/c-demuxSNP-package-demo.html">demuxSNP: supervised demultiplexing using cell hashing and SNPs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/michaelplynch/demultiplexing-bioc23" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Intro to scRNAseq and demultiplexing</h1>
                        <h4 data-toc-skip class="author">Michael
Lynch</h4>
            
            <h4 data-toc-skip class="date">2023-06-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/michaelplynch/demultiplexing-bioc23/blob/HEAD/vignettes/a-intro-to-demultiplexing.Rmd" class="external-link"><code>vignettes/a-intro-to-demultiplexing.Rmd</code></a></small>
      <div class="hidden name"><code>a-intro-to-demultiplexing.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/" class="external-link">viridisLite</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/michaelplynch/demuxSNP" class="external-link">demuxSNP</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CiteFuse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-1"</span>, <span class="st">"0"</span>, <span class="st">"1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction-to-demultiplexing">Introduction to demultiplexing<a class="anchor" aria-label="anchor" href="#introduction-to-demultiplexing"></a>
</h2>
<div class="section level3">
<h3 id="scrnaseq">scRNASeq<a class="anchor" aria-label="anchor" href="#scrnaseq"></a>
</h3>
<p>scRNAseq allows further understanding of cell states, heterogeneity,
in TME and other areas. Sequencing costs are decreasing but remains a
limiting factor in sample size.</p>
<p>For full analysis pipelines see ‘link to OSCA’.</p>
<p>Multiplexing allows cells from multiple biological samples to be
laoded simultaneously on the same lane</p>
</div>
<div class="section level3">
<h3 id="demultiplexing">Demultiplexing<a class="anchor" aria-label="anchor" href="#demultiplexing"></a>
</h3>
<p>Multiplexing is a technique introduced to reduce sequencing cost by
increasing the number of cells loaded per lane, while also allowing
identification of doublets. Multiplexing in scRNAseq involves the
sequencing of samples from different patients, treatment types or
physiological locations together, resulting in significant cost savings.
The cells must then be demultiplexed, or assigned back to their
respective groups. A number of experimental and computational methods
have been proposed to facilitate this, but a universally robust
algorithm remains elusive. Below, we introduce the two main approaches
for demultiplexing and illustrate them with simplified examples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/droplet_encaps.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/droplet_encaps.png" alt="edit to show demultiplexing step" width="301"><p class="caption">
edit to show demultiplexing step
</p>
</div>
<p>Successful demultiplexing is critical, with misassigned or unassigned
cells resulting in increased waste and lower detectable fold changes
between groups.</p>
<div class="section level4">
<h4 id="cell-hashing">Cell Hashing<a class="anchor" aria-label="anchor" href="#cell-hashing"></a>
</h4>
<p>Cells from each group are labelled with a distinct tag (HTO or LMO)
which is sequenced to give a counts matrix. Due to non-specific binding,
these counts form a bimodal distribution. Such methods are generally
computationally efficient. Their classification performance, however, is
highly dependent on the tagging quality and many methods do not account
for uncertainty in classification (<span class="citation">Boggy et al.
(2022)</span>, <span class="citation">Kim et al. (2020)</span> &amp;
<span class="citation">Stoeckius et al. (2018)</span>).</p>
<p>More recent methods, including <a href="https://bioconductor.org/packages/release/bioc/html/demuxmix.html" class="external-link">demuxmix</a>,
assign a probability that a cell is from a particular group, or made up
of multiple groups (doublet). This allows users to define a cut-off
threshold for the assignment confidence. Accounting for uncertainty is
an important feature for these types of algorithms. But, while they give
the user greater flexibility in determining which cells to keep, this
ultimately results in a trade-off between keeping cells which cannot be
confidently called or discarding them - due to issues with tagging
quality rather than RNA quality.</p>
<p>Example matrix:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CITEseq_example"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span>
<span><span class="va">sce_citeseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/preprocessing.html" class="external-link">preprocessing</a></span><span class="op">(</span><span class="va">CITEseq_example</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce_citeseq</span>,<span class="st">"HTO"</span><span class="op">)</span>,<span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="va">hto</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hto</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 4 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##          AAGCCGCGTTGTCTTT GATCGCGGTTATCGGT GGCTGGTAGAGGTTAT CTACACCCAATAGCAA</span></span>
<span><span class="co">## HTO28_5P                .              205                .              562</span></span>
<span><span class="co">## HTO29_5P                2                3                3              209</span></span>
<span><span class="co">## HTO30_5P                1                1                .                .</span></span>
<span><span class="co">## HTO44_5P              548                .              196                .</span></span></code></pre>
<div class="section level5">
<h5 id="overall">Overall:<a class="anchor" aria-label="anchor" href="#overall"></a>
</h5>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">hto</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>,cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,show_column_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-5-1.png" width="100%"></p>
</div>
<div class="section level5">
<h5 id="per-cell">Per cell:<a class="anchor" aria-label="anchor" href="#per-cell"></a>
</h5>
<p>Hashtag counts across cells</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">hto</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">group</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">group</span>, y<span class="op">=</span><span class="va">AAGCCGCGTTGTCTTT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
</div>
<div class="section level5">
<h5 id="per-hashtag">Per hashtag<a class="anchor" aria-label="anchor" href="#per-hashtag"></a>
</h5>
<p>Hashtag counts across a Hashtag (bimodal distriution)</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">bg</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">500</span>,size<span class="op">=</span><span class="fl">5</span>,mu<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sig</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">400</span>,size<span class="op">=</span><span class="fl">10</span>,mu<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bg</span>,<span class="va">sig</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">&lt;-</span><span class="st">"Hashtag1"</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/gghistogram.html" class="external-link">gghistogram</a></span><span class="op">(</span><span class="va">df</span>, x <span class="op">=</span> <span class="st">"Hashtag1"</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html" class="external-link">dittoColors</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, palette <span class="op">=</span> <span class="st">"lancet"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">120</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">750</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">1</span>,bins<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-7-1.png" width="100%"></p>
<p>Reduction in hashing quality leads to reduction in cell hashing
algorithm performance.</p>
</div>
</div>
<div class="section level4">
<h4 id="snps">SNPs<a class="anchor" aria-label="anchor" href="#snps"></a>
</h4>
<p>The second class of methods exploits natural genetic variation
between cells and so can only be used where the groups are genetically
distinct.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/snps.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/snps.png" alt="SNP variation between individuals" width="154"><p class="caption">
SNP variation between individuals
</p>
</div>
<p>Below we describe a simplified example of what SNP data may look
like.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>,<span class="fl">50</span>,<span class="fl">50</span>,<span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">nsnps</span><span class="op">=</span><span class="fl">20</span></span>
<span><span class="va">snps_mat</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">snp</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">nsnps</span>,<span class="fl">1</span>,<span class="fl">0.3</span><span class="op">)</span></span>
<span>  <span class="va">h</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">snp</span>,<span class="va">n</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>,<span class="va">nsnps</span>,<span class="va">n</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">h</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Group"</span>,<span class="va">i</span><span class="op">)</span>,<span class="va">n</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">snps_mat</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">snps_mat</span>,<span class="va">h</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">snps_mat</span><span class="op">[</span><span class="va">snps_mat</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tot_mixed</span><span class="op">&lt;-</span><span class="va">snps_mat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">h1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">snps_mat</span>,cluster_columns<span class="op">=</span><span class="cn">FALSE</span>,cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,col<span class="op">=</span><span class="va">colors</span>,column_split<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">snps_mat</span><span class="op">)</span>,show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,name<span class="op">=</span><span class="st">"SNP status"</span>, row_title <span class="op">=</span> <span class="st">"SNPs"</span>,column_title <span class="op">=</span> <span class="st">"Cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">h1</span>,column_title<span class="op">=</span><span class="st">"Groupwise SNP profiles"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-9-1.png" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">tot_mixed</span>,cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, cluster_rows<span class="op">=</span><span class="cn">FALSE</span>, col<span class="op">=</span><span class="va">colors</span>, show_column_names<span class="op">=</span><span class="cn">FALSE</span>, name<span class="op">=</span><span class="st">"SNP status"</span>, row_title <span class="op">=</span> <span class="st">"SNPs"</span>,column_title <span class="op">=</span> <span class="st">"Cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">h2</span>,column_title<span class="op">=</span><span class="st">"Mixed SNP profiles"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-9-2.png" width="100%"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h3</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">tot_mixed</span>,cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>,cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,row_gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>, column_gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,column_split<span class="op">=</span><span class="fl">4</span>, show_column_names<span class="op">=</span><span class="cn">FALSE</span>, col<span class="op">=</span><span class="va">colors</span>, name<span class="op">=</span><span class="st">"SNP status"</span>, row_title <span class="op">=</span> <span class="st">"SNPs"</span>,column_title <span class="op">=</span> <span class="st">"Cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">h3</span>,column_title<span class="op">=</span><span class="st">"Clustered SNP profiles"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-9-3.png" width="100%"></p>
<p>Using this small example, we can reason how cells pooled from
genetically distinct biological samples might be grouped. In practice
this task is made more complicated due to technical issues such as
sequencing depth, ambient RNA, differing gene expression between cells,
low quality cells, low sequencing coverage, doublets and low quality
cells.</p>
</div>
</div>
<div class="section level3">
<h3 id="sessioninfo">Sessioninfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SeuratObject_4.1.3          Seurat_4.3.0.1             </span></span>
<span><span class="co">##  [3] ggpubr_0.6.0                dittoSeq_1.12.0            </span></span>
<span><span class="co">##  [5] ggplot2_3.4.2               CiteFuse_1.12.0            </span></span>
<span><span class="co">##  [7] demuxSNP_0.99.10            ensembldb_2.24.0           </span></span>
<span><span class="co">##  [9] AnnotationFilter_1.24.0     GenomicFeatures_1.52.1     </span></span>
<span><span class="co">## [11] AnnotationDbi_1.62.1        VariantAnnotation_1.46.0   </span></span>
<span><span class="co">## [13] Rsamtools_2.16.0            Biostrings_2.68.1          </span></span>
<span><span class="co">## [15] XVector_0.40.0              ComplexHeatmap_2.16.0      </span></span>
<span><span class="co">## [17] viridisLite_0.4.2           SingleCellExperiment_1.22.0</span></span>
<span><span class="co">## [19] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">## [21] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [23] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">## [25] BiocGenerics_0.46.0         MatrixGenerics_1.12.2      </span></span>
<span><span class="co">## [27] matrixStats_1.0.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] segmented_1.6-4           spatstat.sparse_3.0-2    </span></span>
<span><span class="co">##   [3] fs_1.6.2                  ProtGenerics_1.32.0      </span></span>
<span><span class="co">##   [5] bitops_1.0-7              httr_1.4.6               </span></span>
<span><span class="co">##   [7] RColorBrewer_1.1-3        doParallel_1.0.17        </span></span>
<span><span class="co">##   [9] ggsci_3.0.0               sctransform_0.3.5        </span></span>
<span><span class="co">##  [11] tools_4.3.1               backports_1.4.1          </span></span>
<span><span class="co">##  [13] utf8_1.2.3                R6_2.5.1                 </span></span>
<span><span class="co">##  [15] lazyeval_0.2.2            uwot_0.1.15              </span></span>
<span><span class="co">##  [17] rhdf5filters_1.12.1       GetoptLong_1.0.5         </span></span>
<span><span class="co">##  [19] sp_2.0-0                  withr_2.5.0              </span></span>
<span><span class="co">##  [21] prettyunits_1.1.1         gridExtra_2.3            </span></span>
<span><span class="co">##  [23] bayesm_3.1-5              progressr_0.13.0         </span></span>
<span><span class="co">##  [25] cli_3.6.1                 textshaping_0.3.6        </span></span>
<span><span class="co">##  [27] spatstat.explore_3.2-1    labeling_0.4.2           </span></span>
<span><span class="co">##  [29] sass_0.4.6                spatstat.data_3.0-1      </span></span>
<span><span class="co">##  [31] robustbase_0.99-0         randomForest_4.7-1.1     </span></span>
<span><span class="co">##  [33] pbapply_1.7-2             ggridges_0.5.4           </span></span>
<span><span class="co">##  [35] pkgdown_2.0.7             systemfonts_1.0.4        </span></span>
<span><span class="co">##  [37] dbscan_1.1-11             parallelly_1.36.0        </span></span>
<span><span class="co">##  [39] BSgenome_1.68.0           limma_3.56.2             </span></span>
<span><span class="co">##  [41] RSQLite_2.3.1             generics_0.1.3           </span></span>
<span><span class="co">##  [43] shape_1.4.6               BiocIO_1.10.0            </span></span>
<span><span class="co">##  [45] combinat_0.0-8            spatstat.random_3.1-5    </span></span>
<span><span class="co">##  [47] ica_1.0-3                 car_3.1-2                </span></span>
<span><span class="co">##  [49] dplyr_1.1.2               Matrix_1.5-4.1           </span></span>
<span><span class="co">##  [51] fansi_1.0.4               abind_1.4-5              </span></span>
<span><span class="co">##  [53] lifecycle_1.0.3           yaml_2.3.7               </span></span>
<span><span class="co">##  [55] edgeR_3.42.4              carData_3.0-5            </span></span>
<span><span class="co">##  [57] rhdf5_2.44.0              BiocFileCache_2.8.0      </span></span>
<span><span class="co">##  [59] Rtsne_0.16                blob_1.2.4               </span></span>
<span><span class="co">##  [61] promises_1.2.0.1          dqrng_0.3.0              </span></span>
<span><span class="co">##  [63] crayon_1.5.2              miniUI_0.1.1.1           </span></span>
<span><span class="co">##  [65] lattice_0.21-8            beachmat_2.16.0          </span></span>
<span><span class="co">##  [67] cowplot_1.1.1             KEGGREST_1.40.0          </span></span>
<span><span class="co">##  [69] demuxmix_1.2.0            magick_2.7.4             </span></span>
<span><span class="co">##  [71] pillar_1.9.0              knitr_1.43               </span></span>
<span><span class="co">##  [73] metapod_1.8.0             rjson_0.2.21             </span></span>
<span><span class="co">##  [75] future.apply_1.11.0       codetools_0.2-19         </span></span>
<span><span class="co">##  [77] leiden_0.4.3              glue_1.6.2               </span></span>
<span><span class="co">##  [79] data.table_1.14.8         vctrs_0.6.3              </span></span>
<span><span class="co">##  [81] png_0.1-8                 gtable_0.3.3             </span></span>
<span><span class="co">##  [83] kernlab_0.9-32            cachem_1.0.8             </span></span>
<span><span class="co">##  [85] xfun_0.39                 mime_0.12                </span></span>
<span><span class="co">##  [87] S4Arrays_1.0.4            tidygraph_1.2.3          </span></span>
<span><span class="co">##  [89] survival_3.5-5            pheatmap_1.0.12          </span></span>
<span><span class="co">##  [91] iterators_1.0.14          statmod_1.5.0            </span></span>
<span><span class="co">##  [93] bluster_1.10.0            ellipsis_0.3.2           </span></span>
<span><span class="co">##  [95] fitdistrplus_1.1-11       ROCR_1.0-11              </span></span>
<span><span class="co">##  [97] nlme_3.1-162              bit64_4.0.5              </span></span>
<span><span class="co">##  [99] progress_1.2.2            filelock_1.0.2           </span></span>
<span><span class="co">## [101] RcppAnnoy_0.0.20          rprojroot_2.0.3          </span></span>
<span><span class="co">## [103] tensorA_0.36.2            bslib_0.5.0              </span></span>
<span><span class="co">## [105] irlba_2.3.5.1             KernSmooth_2.23-21       </span></span>
<span><span class="co">## [107] colorspace_2.1-0          DBI_1.1.3                </span></span>
<span><span class="co">## [109] tidyselect_1.2.0          bit_4.0.5                </span></span>
<span><span class="co">## [111] compiler_4.3.1            curl_5.0.1               </span></span>
<span><span class="co">## [113] compositions_2.0-6        BiocNeighbors_1.18.0     </span></span>
<span><span class="co">## [115] xml2_1.3.4                desc_1.4.2               </span></span>
<span><span class="co">## [117] DelayedArray_0.26.3       plotly_4.10.2            </span></span>
<span><span class="co">## [119] rtracklayer_1.60.0        scales_1.2.1             </span></span>
<span><span class="co">## [121] lmtest_0.9-40             DEoptimR_1.0-14          </span></span>
<span><span class="co">## [123] rappdirs_0.3.3            goftest_1.2-3            </span></span>
<span><span class="co">## [125] stringr_1.5.0             digest_0.6.32            </span></span>
<span><span class="co">## [127] spatstat.utils_3.0-3      mixtools_2.0.0           </span></span>
<span><span class="co">## [129] rmarkdown_2.22            htmltools_0.5.5          </span></span>
<span><span class="co">## [131] pkgconfig_2.0.3           sparseMatrixStats_1.12.1 </span></span>
<span><span class="co">## [133] highr_0.10                dbplyr_2.3.2             </span></span>
<span><span class="co">## [135] fastmap_1.1.1             rlang_1.1.1              </span></span>
<span><span class="co">## [137] GlobalOptions_0.1.2       htmlwidgets_1.6.2        </span></span>
<span><span class="co">## [139] shiny_1.7.4               DelayedMatrixStats_1.22.1</span></span>
<span><span class="co">## [141] farver_2.1.1              jquerylib_0.1.4          </span></span>
<span><span class="co">## [143] zoo_1.8-12                jsonlite_1.8.5           </span></span>
<span><span class="co">## [145] BiocParallel_1.34.2       BiocSingular_1.16.0      </span></span>
<span><span class="co">## [147] RCurl_1.98-1.12           magrittr_2.0.3           </span></span>
<span><span class="co">## [149] scuttle_1.10.1            GenomeInfoDbData_1.2.10  </span></span>
<span><span class="co">## [151] patchwork_1.1.2           Rhdf5lib_1.22.0          </span></span>
<span><span class="co">## [153] munsell_0.5.0             Rcpp_1.0.10              </span></span>
<span><span class="co">## [155] reticulate_1.30           viridis_0.6.3            </span></span>
<span><span class="co">## [157] stringi_1.7.12            ggraph_2.1.0             </span></span>
<span><span class="co">## [159] zlibbioc_1.46.0           MASS_7.3-60              </span></span>
<span><span class="co">## [161] plyr_1.8.8                listenv_0.9.0            </span></span>
<span><span class="co">## [163] parallel_4.3.1            ggrepel_0.9.3            </span></span>
<span><span class="co">## [165] deldir_1.0-9              graphlayouts_1.0.0       </span></span>
<span><span class="co">## [167] splines_4.3.1             tensor_1.5               </span></span>
<span><span class="co">## [169] hms_1.1.3                 circlize_0.4.15          </span></span>
<span><span class="co">## [171] locfit_1.5-9.8            igraph_1.5.0             </span></span>
<span><span class="co">## [173] spatstat.geom_3.2-1       ggsignif_0.6.4           </span></span>
<span><span class="co">## [175] reshape2_1.4.4            biomaRt_2.56.1           </span></span>
<span><span class="co">## [177] ScaledMatrix_1.8.1        XML_3.99-0.14            </span></span>
<span><span class="co">## [179] evaluate_0.21             scran_1.28.1             </span></span>
<span><span class="co">## [181] httpuv_1.6.11             foreach_1.5.2            </span></span>
<span><span class="co">## [183] tweenr_2.0.2              RANN_2.6.1               </span></span>
<span><span class="co">## [185] tidyr_1.3.0               purrr_1.0.1              </span></span>
<span><span class="co">## [187] polyclip_1.10-4           scattermore_1.2          </span></span>
<span><span class="co">## [189] future_1.32.0             clue_0.3-64              </span></span>
<span><span class="co">## [191] ggforce_0.4.1             rsvd_1.0.5               </span></span>
<span><span class="co">## [193] xtable_1.8-4              broom_1.0.5              </span></span>
<span><span class="co">## [195] restfulr_0.0.15           later_1.3.1              </span></span>
<span><span class="co">## [197] rstatix_0.7.2             class_7.3-22             </span></span>
<span><span class="co">## [199] ragg_1.2.5                tibble_3.2.1             </span></span>
<span><span class="co">## [201] memoise_2.0.1             GenomicAlignments_1.36.0 </span></span>
<span><span class="co">## [203] cluster_2.1.4             globals_0.16.2</span></span></code></pre>
</div>
<div class="section level3">
<h3 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-boggy_bff_2022" class="csl-entry">
Boggy, Gregory J, G W McElfresh, Eisa Mahyari, Abigail B Ventura, Scott
G Hansen, Louis J Picker, and Benjamin N Bimber. 2022.
<span>“<span>BFF</span> and <span class="nocase">cellhashR</span>:
Analysis Tools for Accurate Demultiplexing of Cell Hashing Data.”</span>
<em>Bioinformatics</em> 38 (10): 2791–801. <a href="https://doi.org/10.1093/bioinformatics/btac213" class="external-link">https://doi.org/10.1093/bioinformatics/btac213</a>.
</div>
<div id="ref-kim_citefuse_2020" class="csl-entry">
Kim, Hani Jieun, Yingxin Lin, Thomas A Geddes, Jean Yee Hwa Yang, and
Pengyi Yang. 2020. <span>“<span>CiteFuse</span> Enables Multi-Modal
Analysis of <span>CITE</span>-Seq Data.”</span> <em>Bioinformatics</em>
36 (14): 4137–43. <a href="https://doi.org/10.1093/bioinformatics/btaa282" class="external-link">https://doi.org/10.1093/bioinformatics/btaa282</a>.
</div>
<div id="ref-stoeckius_cell_2018" class="csl-entry">
Stoeckius, Marlon, Shiwei Zheng, Brian Houck-Loomis, Stephanie Hao,
Bertrand Z. Yeung, William M. Mauck, Peter Smibert, and Rahul Satija.
2018. <span>“Cell <span>Hashing</span> with Barcoded Antibodies Enables
Multiplexing and Doublet Detection for Single Cell Genomics.”</span>
<em>Genome Biology</em> 19 (1): 224. <a href="https://doi.org/10.1186/s13059-018-1603-1" class="external-link">https://doi.org/10.1186/s13059-018-1603-1</a>.
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Lynch.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
