<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Intro to scRNAseq and demultiplexing • demultiplexingbioc23</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Intro to scRNAseq and demultiplexing">
<meta property="og:description" content="demultiplexingbioc23">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demultiplexingbioc23</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a-intro-to-demultiplexing.html">Intro to scRNAseq and demultiplexing</a>
    </li>
    <li>
      <a href="../articles/b-hashing-benchmarking.html">Simulated benchmarking of hashing based algorithms</a>
    </li>
    <li>
      <a href="../articles/c-demuxSNP-package-demo.html">demuxSNP: supervised demultiplexing using cell hashing and SNPs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/michaelplynch/demultiplexing-bioc23" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Intro to scRNAseq and demultiplexing</h1>
                        <h4 data-toc-skip class="author">Michael
Lynch</h4>
            
            <h4 data-toc-skip class="date">2023-07-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/michaelplynch/demultiplexing-bioc23/blob/HEAD/vignettes/a-intro-to-demultiplexing.Rmd" class="external-link"><code>vignettes/a-intro-to-demultiplexing.Rmd</code></a></small>
      <div class="hidden name"><code>a-intro-to-demultiplexing.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/" class="external-link">viridisLite</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/michaelplynch/demuxSNP" class="external-link">demuxSNP</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CiteFuse</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://michaelplynch.github.io/demultiplexing-bioc23/">demultiplexingbioc23</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-1"</span>, <span class="st">"0"</span>, <span class="st">"1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction-to-demultiplexing">Introduction to demultiplexing<a class="anchor" aria-label="anchor" href="#introduction-to-demultiplexing"></a>
</h2>
<div class="section level3">
<h3 id="scrnaseq">scRNASeq<a class="anchor" aria-label="anchor" href="#scrnaseq"></a>
</h3>
<p>scRNAseq allows further understanding of cell states, heterogeneity,
in TME and other areas. Sequencing costs are decreasing but remains a
limiting factor in sample size.</p>
<p>For full analysis pipelines see <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">‘OSCA’</a> book.</p>
<p>Droplet based single cell sequencing was initially limited by the
doublet rate (a technical artifact where two or more cells are captured
in the same droplet) and the number of cells from each sample a
researched was interested in sequencing. Cells from each biological
sample had to be sequenced on a different lane, resulting in increased
cost.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/nonmultiplexed.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/nonmultiplexed.png" alt="Multiplexed (multi-sample) scRNAseq" width="704"><p class="caption">
Multiplexed (multi-sample) scRNAseq
</p>
</div>
</div>
<div class="section level3">
<h3 id="demultiplexing">Demultiplexing<a class="anchor" aria-label="anchor" href="#demultiplexing"></a>
</h3>
<p>Multiplexing is a technique to reduce sequencing cost by loading
multiple samples per lane. Demultiplexing, is the assignment of cells
back to their original sample, while also allowing the identification of
some doublets cells.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/multiplexing.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/multiplexing.png" alt="Multiplexed (multi-sample) scRNAseq" width="506"><p class="caption">
Multiplexed (multi-sample) scRNAseq
</p>
</div>
<p>A number of experimental and computational methods have been proposed
to facilitate this, but a universally robust algorithm remains elusive.
Below, we introduce the two main approaches for demultiplexing and
illustrate them with simplified examples.</p>
<p>Successful demultiplexing is critical, with misassigned or unassigned
cells resulting in increased waste and lower detectable fold changes
between groups.</p>
<div class="section level4">
<h4 id="cell-hashing">Cell Hashing<a class="anchor" aria-label="anchor" href="#cell-hashing"></a>
</h4>
<p>Cells from each group are labelled with a distinct sequencable
barcode tag (HTO or LMO) which results in a counts matrix after being
sequenced. Due to non-specific binding, these counts form a bimodal
distribution.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/hashing.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/hashing.png" alt="Cell hashing" width="640"><p class="caption">
Cell hashing
</p>
</div>
<p>Such methods are generally computationally efficient. Their
classification performance, however, is highly dependent on the tagging
quality and many methods do not account for uncertainty in
classification (<span class="citation">Boggy et al. (2022)</span>, <span class="citation">Kim et al. (2020)</span> &amp; <span class="citation">Stoeckius et al. (2018)</span>).</p>
<p>More recent methods, including <a href="https://bioconductor.org/packages/release/bioc/html/demuxmix.html" class="external-link">demuxmix</a>,
assign a probability that a cell is from a particular group, or made up
of multiple groups (doublet). This allows users to define a cut-off
threshold for the assignment confidence. Accounting for uncertainty is
an important feature for these types of algorithms. But, while they give
the user greater flexibility in determining which cells to keep, this
ultimately results in a trade-off between keeping cells which cannot be
confidently called or discarding them - due to issues with tagging
quality rather than RNA quality.</p>
<p>Example matrix:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CITEseq_example"</span>, package <span class="op">=</span> <span class="st">"CiteFuse"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/preprocessing.html" class="external-link">preprocessing</a></span><span class="op">(</span><span class="va">CITEseq_example</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>,<span class="st">"HTO"</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>,<span class="st">"HTO"</span><span class="op">)</span>,<span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/normaliseExprs.html" class="external-link">normaliseExprs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">sce</span>, altExp_name <span class="op">=</span> <span class="st">"HTO"</span>, transform <span class="op">=</span> <span class="st">"log"</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/crossSampleDoublets.html" class="external-link">crossSampleDoublets</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## number of iterations= 20 </span></span>
<span><span class="co">## number of iterations= 24 </span></span>
<span><span class="co">## number of iterations= 46 </span></span>
<span><span class="co">## number of iterations= 50</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hto</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>,<span class="st">"HTO"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hto</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 4 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##          AAGCCGCGTTGTCTTT GATCGCGGTTATCGGT GGCTGGTAGAGGTTAT CTACACCCAATAGCAA</span></span>
<span><span class="co">## HTO28_5P                .              205                .              562</span></span>
<span><span class="co">## HTO29_5P                2                3                3              209</span></span>
<span><span class="co">## HTO30_5P                1                1                .                .</span></span>
<span><span class="co">## HTO44_5P              548                .              196                .</span></span></code></pre>
<div class="section level5">
<h5 id="overall">Overall:<a class="anchor" aria-label="anchor" href="#overall"></a>
</h5>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">hto</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>,cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,show_column_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
</div>
<div class="section level5">
<h5 id="per-cell">Per cell:<a class="anchor" aria-label="anchor" href="#per-cell"></a>
</h5>
<p>Hashtag counts across cells</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">hto</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">group</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">group</span>, y<span class="op">=</span><span class="va">AAGCCGCGTTGTCTTT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-9-1.png" width="100%"></p>
</div>
<div class="section level5">
<h5 id="per-hashtag">Per hashtag<a class="anchor" aria-label="anchor" href="#per-hashtag"></a>
</h5>
<p>Hashtag counts across a Hashtag (bimodal distriution)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">bg</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">500</span>,size<span class="op">=</span><span class="fl">5</span>,mu<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sig</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">400</span>,size<span class="op">=</span><span class="fl">10</span>,mu<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>,<span class="st">"HTO"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">myplots</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/plot_hashtag.html">plot_hashtag</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Using `bins = 30` by default. Pick better value with the argument</span></span>
<span><span class="co">## `bins`.</span></span>
<span></span>
<span><span class="co">## Warning: Using `bins = 30` by default. Pick better value with the argument</span></span>
<span><span class="co">## `bins`.</span></span>
<span></span>
<span><span class="co">## Warning: Using `bins = 30` by default. Pick better value with the argument</span></span>
<span><span class="co">## `bins`.</span></span>
<span></span>
<span><span class="co">## Warning: Using `bins = 30` by default. Pick better value with the argument</span></span>
<span><span class="co">## `bins`.</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>,<span class="va">myplots</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 1 rows containing missing values (`geom_bar()`).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 1 rows containing missing values (`geom_bar()`).</span></span>
<span><span class="co">## Removed 1 rows containing missing values (`geom_bar()`).</span></span>
<span><span class="co">## Removed 1 rows containing missing values (`geom_bar()`).</span></span></code></pre>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-10-1.png" width="100%"></p>
<p>Reduction in hashing quality leads to reduction in cell hashing
algorithm performance.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, altexp<span class="op">=</span><span class="st">"HTO"</span>, exprs_values <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in check_numbers(k = k, nu = nu, nv = nv, limit = min(dim(x)) - : more</span></span>
<span><span class="co">## singular values/vectors requested than available</span></span></code></pre>
<pre><code><span><span class="co">## Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =</span></span>
<span><span class="co">## TRUE, : You're computing too large a percentage of total singular values, use a</span></span>
<span><span class="co">## standard svd instead.</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoDimPlot.html" class="external-link">dittoDimPlot</a></span><span class="op">(</span><span class="va">sce</span>,reduction.use<span class="op">=</span><span class="st">"PCA"</span>,<span class="st">"doubletClassify_between_label"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, altexp <span class="op">=</span> <span class="st">"HTO"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoDimPlot.html" class="external-link">dittoDimPlot</a></span><span class="op">(</span><span class="va">sce</span>,reduction.use<span class="op">=</span><span class="st">"TSNE"</span>,<span class="st">"doubletClassify_between_label"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-11-2.png" width="700"></p>
</div>
</div>
<div class="section level4">
<h4 id="snps">SNPs<a class="anchor" aria-label="anchor" href="#snps"></a>
</h4>
<p>The second class of methods exploits natural genetic variation
between cells and so can only be used where the groups are genetically
distinct.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/snps.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/snps.png" alt="SNP variation between individuals" width="50%"><p class="caption">
SNP variation between individuals
</p>
</div>
<div class="section level5">
<h5 id="sample-simulated-snps">Sample simulated SNPs<a class="anchor" aria-label="anchor" href="#sample-simulated-snps"></a>
</h5>
<p>Below we describe a simplified example of what SNP data may look
like.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>,<span class="fl">50</span>,<span class="fl">50</span>,<span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">nsnps</span><span class="op">=</span><span class="fl">20</span></span>
<span><span class="va">snps_mat</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">snp</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">nsnps</span>,<span class="fl">1</span>,<span class="fl">0.3</span><span class="op">)</span></span>
<span>  <span class="va">h</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">snp</span>,<span class="va">n</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>,<span class="va">nsnps</span>,<span class="va">n</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">h</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Group"</span>,<span class="va">i</span><span class="op">)</span>,<span class="va">n</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">snps_mat</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">snps_mat</span>,<span class="va">h</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">snps_mat</span><span class="op">[</span><span class="va">snps_mat</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tot_mixed</span><span class="op">&lt;-</span><span class="va">snps_mat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">h1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">snps_mat</span>,</span>
<span>            cluster_columns<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>            cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>            col<span class="op">=</span><span class="va">colors</span>,</span>
<span>            column_split<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">snps_mat</span><span class="op">)</span>,</span>
<span>            show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            name<span class="op">=</span><span class="st">"SNP status"</span>, </span>
<span>            row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>            column_title <span class="op">=</span> <span class="st">"Cells"</span>,</span>
<span>            heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            title <span class="op">=</span> <span class="st">"SNP status"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>            labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Absent"</span>, <span class="st">"No reads"</span>, <span class="st">"Present"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">h1</span>,column_title<span class="op">=</span><span class="st">"Groupwise SNP profiles"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-13-1.png" width="100%"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">tot_mixed</span>,</span>
<span>            cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>            cluster_rows<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>            col<span class="op">=</span><span class="va">colors</span>, </span>
<span>            show_column_names<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>            name<span class="op">=</span><span class="st">"SNP status"</span>, </span>
<span>            row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>            column_title <span class="op">=</span> <span class="st">"Cells"</span>,</span>
<span>            heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            title <span class="op">=</span> <span class="st">"SNP status"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>            labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Absent"</span>, <span class="st">"No reads"</span>, <span class="st">"Present"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">h2</span>,column_title<span class="op">=</span><span class="st">"Mixed SNP profiles"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-13-2.png" width="100%"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h3</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">tot_mixed</span>,</span>
<span>            cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>            row_gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>, </span>
<span>            column_gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>            column_split<span class="op">=</span><span class="fl">4</span>, </span>
<span>            show_column_names<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>            col<span class="op">=</span><span class="va">colors</span>, name<span class="op">=</span><span class="st">"SNP status"</span>, </span>
<span>            row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>            column_title <span class="op">=</span> <span class="st">"Cells"</span>,</span>
<span>            heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            title <span class="op">=</span> <span class="st">"SNP status"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>            labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Absent"</span>, <span class="st">"No reads"</span>, <span class="st">"Present"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">h3</span>,column_title<span class="op">=</span><span class="st">"Clustered SNP profiles"</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-13-3.png" width="100%"></p>
<p>Using this small example, we can reason how cells pooled from
genetically distinct biological samples might be grouped. In practice
this task is made more complicated due to technical issues such as
sequencing depth, ambient RNA, differing gene expression between cells,
low quality cells, low sequencing coverage, doublets and low quality
cells.</p>
</div>
<div class="section level5">
<h5 id="snp-profiles">SNP profiles<a class="anchor" aria-label="anchor" href="#snp-profiles"></a>
</h5>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">vartrix_consensus_snps</span>,package<span class="op">=</span><span class="st">'demuxSNP'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">vartrix_consensus_snps</span>,</span>
<span>        cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>        show_column_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"SNP status"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, </span>
<span>        labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No reads"</span>, <span class="st">"Ref"</span>, <span class="st">"Alt"</span>,<span class="st">"Both"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `use_raster` is automatically set to TRUE for a matrix with more than</span></span>
<span><span class="co">## 2000 rows. You can control `use_raster` argument by explicitly setting</span></span>
<span><span class="co">## TRUE/FALSE to it.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Set `ht_opt$message = FALSE` to turn off this message.</span></span></code></pre>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Only some of these SNPs are present in the six samples, so lets
filtered out some of the SNPs we don’t have reads for across the
board.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered_mat</span><span class="op">&lt;-</span><span class="va">vartrix_consensus_snps</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">vartrix_consensus_snps</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">1000</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mode.html" class="external-link">mode</a></span><span class="op">(</span><span class="va">filtered_mat</span><span class="op">)</span><span class="op">&lt;-</span><span class="st">"integer"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">filtered_mat</span>,</span>
<span>        cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>        show_column_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"SNP status"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, </span>
<span>        labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No reads"</span>, <span class="st">"Ref"</span>, <span class="st">"Alt"</span>,<span class="st">"Both"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered_mat</span><span class="op">[</span><span class="va">filtered_mat</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">filtered_mat</span><span class="op">[</span><span class="va">filtered_mat</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">filtered_mat</span><span class="op">[</span><span class="va">filtered_mat</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">filtered_mat</span><span class="op">[</span><span class="va">filtered_mat</span> <span class="op">==</span> <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></code></pre></div>
<p>Finally, we simplify the data to present, absent, or no reads in
relation to our six samples.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">filtered_mat</span>,</span>
<span>        cluster_rows<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>        show_column_dend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        col <span class="op">=</span> <span class="va">colors</span>,</span>
<span>        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"SNP status"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>        labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Absent"</span>,<span class="st">"No reads"</span>, <span class="st">"Present"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="a-intro-to-demultiplexing_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="sessioninfo">Sessioninfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h3>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] demultiplexingbioc23_0.99.0 scater_1.28.0              </span></span>
<span><span class="co">##  [3] scuttle_1.10.1              gridExtra_2.3              </span></span>
<span><span class="co">##  [5] SeuratObject_4.1.3          Seurat_4.3.0.1             </span></span>
<span><span class="co">##  [7] ggpubr_0.6.0                dittoSeq_1.12.0            </span></span>
<span><span class="co">##  [9] ggplot2_3.4.2               CiteFuse_1.12.0            </span></span>
<span><span class="co">## [11] demuxSNP_0.99.15            ensembldb_2.24.0           </span></span>
<span><span class="co">## [13] AnnotationFilter_1.24.0     GenomicFeatures_1.52.1     </span></span>
<span><span class="co">## [15] AnnotationDbi_1.62.2        VariantAnnotation_1.46.0   </span></span>
<span><span class="co">## [17] Rsamtools_2.16.0            Biostrings_2.68.1          </span></span>
<span><span class="co">## [19] XVector_0.40.0              ComplexHeatmap_2.16.0      </span></span>
<span><span class="co">## [21] viridisLite_0.4.2           SingleCellExperiment_1.22.0</span></span>
<span><span class="co">## [23] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">## [25] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [27] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">## [29] BiocGenerics_0.46.0         MatrixGenerics_1.12.2      </span></span>
<span><span class="co">## [31] matrixStats_1.0.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] progress_1.2.2            goftest_1.2-3            </span></span>
<span><span class="co">##   [3] vctrs_0.6.3               spatstat.random_3.1-5    </span></span>
<span><span class="co">##   [5] digest_0.6.33             png_0.1-8                </span></span>
<span><span class="co">##   [7] shape_1.4.6               ggrepel_0.9.3            </span></span>
<span><span class="co">##   [9] deldir_1.0-9              parallelly_1.36.0        </span></span>
<span><span class="co">##  [11] magick_2.7.4              MASS_7.3-60              </span></span>
<span><span class="co">##  [13] pkgdown_2.0.7             reshape2_1.4.4           </span></span>
<span><span class="co">##  [15] foreach_1.5.2             httpuv_1.6.11            </span></span>
<span><span class="co">##  [17] withr_2.5.0               xfun_0.39                </span></span>
<span><span class="co">##  [19] ellipsis_0.3.2            survival_3.5-5           </span></span>
<span><span class="co">##  [21] memoise_2.0.1             ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##  [23] ggsci_3.0.0               mixtools_2.0.0           </span></span>
<span><span class="co">##  [25] systemfonts_1.0.4         ragg_1.2.5               </span></span>
<span><span class="co">##  [27] zoo_1.8-12                GlobalOptions_0.1.2      </span></span>
<span><span class="co">##  [29] pbapply_1.7-2             DEoptimR_1.1-0           </span></span>
<span><span class="co">##  [31] prettyunits_1.1.1         KEGGREST_1.40.0          </span></span>
<span><span class="co">##  [33] promises_1.2.0.1          httr_1.4.6               </span></span>
<span><span class="co">##  [35] rstatix_0.7.2             restfulr_0.0.15          </span></span>
<span><span class="co">##  [37] globals_0.16.2            fitdistrplus_1.1-11      </span></span>
<span><span class="co">##  [39] rhdf5filters_1.12.1       demuxmix_1.2.0           </span></span>
<span><span class="co">##  [41] rhdf5_2.44.0              miniUI_0.1.1.1           </span></span>
<span><span class="co">##  [43] generics_0.1.3            curl_5.0.1               </span></span>
<span><span class="co">##  [45] zlibbioc_1.46.0           ScaledMatrix_1.8.1       </span></span>
<span><span class="co">##  [47] ggraph_2.1.0              polyclip_1.10-4          </span></span>
<span><span class="co">##  [49] randomForest_4.7-1.1      GenomeInfoDbData_1.2.10  </span></span>
<span><span class="co">##  [51] xtable_1.8-4              stringr_1.5.0            </span></span>
<span><span class="co">##  [53] desc_1.4.2                doParallel_1.0.17        </span></span>
<span><span class="co">##  [55] evaluate_0.21             S4Arrays_1.0.5           </span></span>
<span><span class="co">##  [57] BiocFileCache_2.8.0       hms_1.1.3                </span></span>
<span><span class="co">##  [59] irlba_2.3.5.1             colorspace_2.1-0         </span></span>
<span><span class="co">##  [61] filelock_1.0.2            ROCR_1.0-11              </span></span>
<span><span class="co">##  [63] reticulate_1.30           spatstat.data_3.0-1      </span></span>
<span><span class="co">##  [65] magrittr_2.0.3            lmtest_0.9-40            </span></span>
<span><span class="co">##  [67] later_1.3.1               viridis_0.6.4            </span></span>
<span><span class="co">##  [69] lattice_0.21-8            spatstat.geom_3.2-4      </span></span>
<span><span class="co">##  [71] future.apply_1.11.0       robustbase_0.99-0        </span></span>
<span><span class="co">##  [73] scattermore_1.2           XML_3.99-0.14            </span></span>
<span><span class="co">##  [75] cowplot_1.1.1             RcppAnnoy_0.0.21         </span></span>
<span><span class="co">##  [77] class_7.3-22              pillar_1.9.0             </span></span>
<span><span class="co">##  [79] nlme_3.1-162              iterators_1.0.14         </span></span>
<span><span class="co">##  [81] compiler_4.3.1            beachmat_2.16.0          </span></span>
<span><span class="co">##  [83] stringi_1.7.12            tensor_1.5               </span></span>
<span><span class="co">##  [85] GenomicAlignments_1.36.0  plyr_1.8.8               </span></span>
<span><span class="co">##  [87] crayon_1.5.2              abind_1.4-5              </span></span>
<span><span class="co">##  [89] BiocIO_1.10.0             locfit_1.5-9.8           </span></span>
<span><span class="co">##  [91] sp_2.0-0                  graphlayouts_1.0.0       </span></span>
<span><span class="co">##  [93] bit_4.0.5                 dplyr_1.1.2              </span></span>
<span><span class="co">##  [95] codetools_0.2-19          textshaping_0.3.6        </span></span>
<span><span class="co">##  [97] BiocSingular_1.16.0       bslib_0.5.0              </span></span>
<span><span class="co">##  [99] GetoptLong_1.0.5          plotly_4.10.2            </span></span>
<span><span class="co">## [101] mime_0.12                 splines_4.3.1            </span></span>
<span><span class="co">## [103] circlize_0.4.15           Rcpp_1.0.11              </span></span>
<span><span class="co">## [105] dbplyr_2.3.3              sparseMatrixStats_1.12.2 </span></span>
<span><span class="co">## [107] knitr_1.43                blob_1.2.4               </span></span>
<span><span class="co">## [109] utf8_1.2.3                clue_0.3-64              </span></span>
<span><span class="co">## [111] fs_1.6.3                  listenv_0.9.0            </span></span>
<span><span class="co">## [113] DelayedMatrixStats_1.22.1 ggsignif_0.6.4           </span></span>
<span><span class="co">## [115] tibble_3.2.1              Matrix_1.6-0             </span></span>
<span><span class="co">## [117] statmod_1.5.0             tweenr_2.0.2             </span></span>
<span><span class="co">## [119] pkgconfig_2.0.3           pheatmap_1.0.12          </span></span>
<span><span class="co">## [121] tools_4.3.1               cachem_1.0.8             </span></span>
<span><span class="co">## [123] RSQLite_2.3.1             DBI_1.1.3                </span></span>
<span><span class="co">## [125] fastmap_1.1.1             rmarkdown_2.23           </span></span>
<span><span class="co">## [127] scales_1.2.1              ica_1.0-3                </span></span>
<span><span class="co">## [129] broom_1.0.5               sass_0.4.7               </span></span>
<span><span class="co">## [131] patchwork_1.1.2           carData_3.0-5            </span></span>
<span><span class="co">## [133] RANN_2.6.1                farver_2.1.1             </span></span>
<span><span class="co">## [135] tidygraph_1.2.3           yaml_2.3.7               </span></span>
<span><span class="co">## [137] bayesm_3.1-5              rtracklayer_1.60.0       </span></span>
<span><span class="co">## [139] cli_3.6.1                 purrr_1.0.1              </span></span>
<span><span class="co">## [141] leiden_0.4.3              lifecycle_1.0.3          </span></span>
<span><span class="co">## [143] dbscan_1.1-11             uwot_0.1.16              </span></span>
<span><span class="co">## [145] bluster_1.10.0            kernlab_0.9-32           </span></span>
<span><span class="co">## [147] backports_1.4.1           BiocParallel_1.34.2      </span></span>
<span><span class="co">## [149] gtable_0.3.3              rjson_0.2.21             </span></span>
<span><span class="co">## [151] ggridges_0.5.4            progressr_0.13.0         </span></span>
<span><span class="co">## [153] parallel_4.3.1            limma_3.56.2             </span></span>
<span><span class="co">## [155] jsonlite_1.8.7            edgeR_3.42.4             </span></span>
<span><span class="co">## [157] bitops_1.0-7              bit64_4.0.5              </span></span>
<span><span class="co">## [159] Rtsne_0.16                spatstat.utils_3.0-3     </span></span>
<span><span class="co">## [161] BiocNeighbors_1.18.0      jquerylib_0.1.4          </span></span>
<span><span class="co">## [163] highr_0.10                metapod_1.8.0            </span></span>
<span><span class="co">## [165] dqrng_0.3.0               segmented_1.6-4          </span></span>
<span><span class="co">## [167] lazyeval_0.2.2            shiny_1.7.4.1            </span></span>
<span><span class="co">## [169] htmltools_0.5.5           sctransform_0.3.5        </span></span>
<span><span class="co">## [171] rappdirs_0.3.3            glue_1.6.2               </span></span>
<span><span class="co">## [173] RCurl_1.98-1.12           rprojroot_2.0.3          </span></span>
<span><span class="co">## [175] scran_1.28.2              BSgenome_1.68.0          </span></span>
<span><span class="co">## [177] igraph_1.5.0.1            R6_2.5.1                 </span></span>
<span><span class="co">## [179] tidyr_1.3.0               labeling_0.4.2           </span></span>
<span><span class="co">## [181] cluster_2.1.4             Rhdf5lib_1.22.0          </span></span>
<span><span class="co">## [183] compositions_2.0-6        DelayedArray_0.26.7      </span></span>
<span><span class="co">## [185] tidyselect_1.2.0          vipor_0.4.5              </span></span>
<span><span class="co">## [187] ProtGenerics_1.32.0       tensorA_0.36.2           </span></span>
<span><span class="co">## [189] ggforce_0.4.1             xml2_1.3.5               </span></span>
<span><span class="co">## [191] car_3.1-2                 future_1.33.0            </span></span>
<span><span class="co">## [193] rsvd_1.0.5                munsell_0.5.0            </span></span>
<span><span class="co">## [195] KernSmooth_2.23-21        data.table_1.14.8        </span></span>
<span><span class="co">## [197] htmlwidgets_1.6.2         RColorBrewer_1.1-3       </span></span>
<span><span class="co">## [199] biomaRt_2.56.1            rlang_1.1.1              </span></span>
<span><span class="co">## [201] spatstat.sparse_3.0-2     spatstat.explore_3.2-1   </span></span>
<span><span class="co">## [203] fansi_1.0.4               Cairo_1.6-0              </span></span>
<span><span class="co">## [205] beeswarm_0.4.0</span></span></code></pre>
</div>
<div class="section level3">
<h3 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-boggy_bff_2022" class="csl-entry">
Boggy, Gregory J, G W McElfresh, Eisa Mahyari, Abigail B Ventura, Scott
G Hansen, Louis J Picker, and Benjamin N Bimber. 2022.
<span>“<span>BFF</span> and <span class="nocase">cellhashR</span>:
Analysis Tools for Accurate Demultiplexing of Cell Hashing Data.”</span>
<em>Bioinformatics</em> 38 (10): 2791–801. <a href="https://doi.org/10.1093/bioinformatics/btac213" class="external-link">https://doi.org/10.1093/bioinformatics/btac213</a>.
</div>
<div id="ref-kim_citefuse_2020" class="csl-entry">
Kim, Hani Jieun, Yingxin Lin, Thomas A Geddes, Jean Yee Hwa Yang, and
Pengyi Yang. 2020. <span>“<span>CiteFuse</span> Enables Multi-Modal
Analysis of <span>CITE</span>-Seq Data.”</span> <em>Bioinformatics</em>
36 (14): 4137–43. <a href="https://doi.org/10.1093/bioinformatics/btaa282" class="external-link">https://doi.org/10.1093/bioinformatics/btaa282</a>.
</div>
<div id="ref-stoeckius_cell_2018" class="csl-entry">
Stoeckius, Marlon, Shiwei Zheng, Brian Houck-Loomis, Stephanie Hao,
Bertrand Z. Yeung, William M. Mauck, Peter Smibert, and Rahul Satija.
2018. <span>“Cell <span>Hashing</span> with Barcoded Antibodies Enables
Multiplexing and Doublet Detection for Single Cell Genomics.”</span>
<em>Genome Biology</em> 19 (1): 224. <a href="https://doi.org/10.1186/s13059-018-1603-1" class="external-link">https://doi.org/10.1186/s13059-018-1603-1</a>.
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Lynch.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
