<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>demuxSNP: supervised demultiplexing using cell hashing and SNPs • demultiplexingbioc23</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="demuxSNP: supervised demultiplexing using cell hashing and SNPs">
<meta property="og:description" content="demultiplexingbioc23">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demultiplexingbioc23</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a-intro-to-demultiplexing.html">Intro to scRNAseq and demultiplexing</a>
    </li>
    <li>
      <a href="../articles/b-hashing-benchmarking.html">Simulated benchmarking of hashing based algorithms</a>
    </li>
    <li>
      <a href="../articles/c-demuxSNP-package-demo.html">demuxSNP: supervised demultiplexing using cell hashing and SNPs</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/michaelplynch/demultiplexing-bioc23" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="c-demuxSNP-package-demo_files/kePrint-0.0.1/kePrint.js"></script><link href="c-demuxSNP-package-demo_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>demuxSNP: supervised demultiplexing using cell
hashing and SNPs</h1>
                        <h4 data-toc-skip class="author">Michael
Lynch</h4>
            <address class="author_afil">
      University of
Limerick<br><h4 data-toc-skip class="author">Aedin
Culhane</h4>
            <address class="author_afil">
      University of Limerick<br><small class="dont-index">Source: <a href="https://github.com/michaelplynch/demultiplexing-bioc23/blob/HEAD/vignettes/c-demuxSNP-package-demo.Rmd" class="external-link"><code>vignettes/c-demuxSNP-package-demo.Rmd</code></a></small>
      <div class="hidden name"><code>c-demuxSNP-package-demo.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/michaelplynch/demuxSNP" class="external-link">demuxSNP</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/" class="external-link">viridisLite</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">utils</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-1"</span>, <span class="st">"0"</span>, <span class="st">"1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>To refresh, demultiplexing allows reduction of sequencing costs by
sequencing cells from multiple biological sampels on the same lane, as
well as allowing identifying of the increased proportion of doublets
associated with greater cells loaded. Poor demultiplexing performance
results in waste in valuable data, reduction in LFC used in DE or
otherwise potentially confusing results.</p>
<div class="section level3">
<h3 id="existing-methods">Existing Methods<a class="anchor" aria-label="anchor" href="#existing-methods"></a>
</h3>
<p>Limitations of existing methods:</p>
<p>Hashing algorithms can correctly call some but not all cells
correctly and is highly dependent on the hashing quality</p>
<p>SNPs algorithms either: Genotype based methods incur additoinal cost
(Demuxlet) Genotype-free methods struggle to identify rarer samples (due
to lack of genotype information)</p>
</div>
<div class="section level3">
<h3 id="demuxsnp-motivation">demuxSNP Motivation<a class="anchor" aria-label="anchor" href="#demuxsnp-motivation"></a>
</h3>
<p><strong>With cell hashing, we can confidently demultiplex
<em>some</em> but not <em>all</em> cells. Using these high confidence
cells, we can learn the SNPs associated with each group. This SNP
information can then be used to assign remaining cells (which we could
not confidently call using cell hashing) to their most similar group
based on their SNP profile.</strong></p>
<p>Novel features:</p>
<ul>
<li><p>Uses both cell hashing and SNP data. Current methods are limited
to using one or the other.</p></li>
<li><p>Selects SNPs based on being located in a gene expressed in a
large proportion of cells to reduce noise, computational cost and
increase interpretability.</p></li>
</ul>
<p>Impact:</p>
<ul>
<li>Users can visually confirm validity (or lack thereof) of existing
demultiplexing results in a tangible manner.</li>
<li>Users can recover otherwise high quality cells which could not be
confidently assigned using other methods.</li>
<li>Cells from groups which are present in lower proportions may be
classified better than with unsupervised SNP approaches.</li>
</ul>
<p>Note: the approach used here differs from most SNP methods in that it
is supervised. We attain knowledge of which SNPs are associated with
which patients from the high confidence cells then use this to train a
classifier. It is similar to demuxlet in the sense that the classifier
uses group specific SNP information, <strong>however</strong> our method
does not require the expense of genotyping and so may be much more
widely applicable.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##biocdevel</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"demuxSNP"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## or for development version:</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"michaelplynch/demuxSNP"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quick-usage">Quick Usage<a class="anchor" aria-label="anchor" href="#quick-usage"></a>
</h3>
<p>For full list of arguments and explanation of each function, please
refer to relevant documentation.</p>
</div>
<div class="section level3">
<h3 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h3>
<p>? maybe remove this for demo</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Data loading</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>, </span>
<span>     <span class="va">commonvariants_1kgenomes_subset</span>, </span>
<span>     <span class="va">vartrix_consensus_snps</span>, </span>
<span>     package <span class="op">=</span> <span class="st">"demuxSNP"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">small_sce</span><span class="op">&lt;-</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="va">ensdb</span> <span class="op">&lt;-</span> <span class="fu">EnsDb.Hsapiens.v86</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/EnsDb.Hsapiens.v86/man/package.html" class="external-link">EnsDb.Hsapiens.v86</a></span></span>
<span></span>
<span><span class="co">#Preprocessing</span></span>
<span><span class="va">top_genes</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/common_genes.html" class="external-link">common_genes</a></span><span class="op">(</span><span class="va">small_sce</span><span class="op">)</span></span>
<span><span class="va">vcf_sub</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/subset_vcf.html" class="external-link">subset_vcf</a></span><span class="op">(</span><span class="va">commonvariants_1kgenomes_subset</span>, <span class="va">top_genes</span>, <span class="va">ensdb</span><span class="op">)</span></span>
<span><span class="va">small_sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/high_conf_calls.html" class="external-link">high_conf_calls</a></span><span class="op">(</span><span class="va">small_sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Use subsetted vcf with VarTrix in default 'consensus' mode to generate SNPs </span></span>
<span><span class="co">#matrix</span></span>
<span></span>
<span><span class="va">small_sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/add_snps.html" class="external-link">add_snps</a></span><span class="op">(</span><span class="va">small_sce</span>,<span class="va">vartrix_consensus_snps</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">small_sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/reassign.html" class="external-link">reassign</a></span><span class="op">(</span><span class="va">small_sce</span>,k<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">small_sce</span><span class="op">$</span><span class="va">knn</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Doublet Hashtag1 Hashtag3 Hashtag4 Hashtag5 Hashtag6 </span></span>
<span><span class="co">#&gt;       30       11       16        7        4       32</span></span></code></pre></div>
<div class="section level4">
<h4 id="function-explanation">Function Explanation<a class="anchor" aria-label="anchor" href="#function-explanation"></a>
</h4>
<p><code>top_genes &lt;- common_genes(sce)</code>: Returns the genes
which are expressed (expression &gt; 0) in the highest proportion of
cells. These genes are used below to subset the .vcf file.</p>
<p><code>new_vcf &lt;- subset_vcf(vcf = vcf, top_genes = top_genes, ensdb = ensdb)</code>:
Subsets a supplied .vcf to SNP locations within the genes supplied. The
ranges of the genes are extracted from the EnsDb object.</p>
<p><code>sce &lt;- high_conf_calls(sce = sce, assay = "HTO")</code>:
Takes a SingleCellExperiment object with HTO altExp, runs <a href="https://bioconductor.org/packages/release/bioc/html/demuxmix.html" class="external-link">demuxmix</a>
and returns a factor of assigned labels, a logical vector indicating
high confidence calls and a logical vector indicating which cells to
predict (all).</p>
<p><code>sce &lt;- add_snps(sce = sce, mat = vartrix_consensus_snps, thresh = 0.8)</code>:
Adds the SNP data from VarTrix (default consensus mode) to the
SingleCellExperiment object as an altExp. Additionally, filters out SNPs
with no reads in less than ‘thresh’ proportion of cells.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/reassign.html" class="external-link">reassign</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    k <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    d <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    train_cells <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">train</span>,</span>
<span>    predict_cells <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">predict</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Reassigns cells based on SNP profiles of high confidence cells.
Singlet training data is based on high confidence singlet assignment.
Doublets are simulated by systematically sampling and combining the SNP
profiles of n cells pairs from each grouping combination. Cells to be
used for training data are specified by “train_cells” (logical). Cells
to be used for prediction are specified by “predict_cells” (logical),
this may also include the training data.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="exploratory-analysis">Exploratory Analysis<a class="anchor" aria-label="anchor" href="#exploratory-analysis"></a>
</h2>
<p>The sample dataset is a subset of ccRCC scRNAseq with six
patients.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"./images/ccrcc-samples.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/ccrcc-samples.png" alt="Multiplexed scRNASeq sample dataset" width="618"><p class="caption">
Multiplexed scRNASeq sample dataset
</p>
</div>
<p>We load three data objects: a SingleCellExperiment object containing
RNA and HTO counts, a .vcf file of class CollapsedVCF containing SNPs
from 1000 Genomes common variants and a matrix containing SNP
information for each cell (we will show you how to generate this SNPs
matrix using <a href="https://github.com/10XGenomics/vartrix" class="external-link">VarTrix</a> outside of R).
We have already removed low quality cells (library size&lt;1,000 and
percentage of genes mapping to mitochondrial genes&gt;10%).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">commonvariants_1kgenomes_subset</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "CollapsedVCF"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "VariantAnnotation"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">vartrix_consensus_snps</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "matrix" "array"</span></span></code></pre></div>
<p>The HTO or LMO distribution is usually bimodal, with a signal (high
counts) and background distribution (low counts) caused by non-specific
binding. Ideally, these distributions would be clearly separated with no
overlap, but in practice, this is not always the case. In our example
data, we see that the signal and noise overlap to varying extents in
each group.</p>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-9-1.png" width="816"></p>
<p>We will begin by running Seurat’s HTODemux, a popular HTO
demultiplexing algorithm on the data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">)</span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">)</span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/HTODemux.html" class="external-link">HTODemux</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">)</span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/MULTIseqDemux.html" class="external-link">MULTIseqDemux</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">)</span></span>
<span><span class="va">seurat</span><span class="op">$</span><span class="va">hash.ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">$</span><span class="va">hash.ID</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seurat</span><span class="op">$</span><span class="va">MULTI_ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">$</span><span class="va">MULTI_ID</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="va">seurat</span><span class="op">$</span><span class="va">hash.ID</span></span>
<span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">multiseq</span><span class="op">&lt;-</span> <span class="va">seurat</span><span class="op">$</span><span class="va">MULTI_ID</span></span>
<span></span>
<span><span class="va">t</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">seurat</span><span class="op">)</span></span>
<span><span class="va">u</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kable_styling</a></span><span class="op">(</span>bootstrap_options <span class="op">=</span> <span class="st">"bordered"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered" style="margin-left: auto; margin-right: auto;"><tbody>
<tr>
<td style="text-align:left;">
Doublet
</td>
<td style="text-align:right;">
633
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag1
</td>
<td style="text-align:right;">
121
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag2
</td>
<td style="text-align:right;">
29
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag3
</td>
<td style="text-align:right;">
264
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag4
</td>
<td style="text-align:right;">
158
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag5
</td>
<td style="text-align:right;">
177
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag6
</td>
<td style="text-align:right;">
383
</td>
</tr>
<tr>
<td style="text-align:left;">
Negative
</td>
<td style="text-align:right;">
235
</td>
</tr>
</tbody></table>
<p>Although HTO library size of the Negative group is low, the RNA
library size is similar to that of other groups, indicating that they
may be misclassified as ‘Negative’ due to their tagging quality rather
than overall RNA quality.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoPlot.html" class="external-link">dittoPlot</a></span><span class="op">(</span><span class="va">seurat</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">$</span><span class="va">nCount_HTO</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"hash.ID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoPlot.html" class="external-link">dittoPlot</a></span><span class="op">(</span><span class="va">seurat</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"hash.ID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-11-2.png" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CiteFuse</span><span class="op">)</span></span>
<span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/CiteFuse/man/crossSampleDoublets.html" class="external-link">crossSampleDoublets</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; number of iterations= 79 </span></span>
<span><span class="co">#&gt; number of iterations= 60 </span></span>
<span><span class="co">#&gt; number of iterations= 72 </span></span>
<span><span class="co">#&gt; number of iterations= 80 </span></span>
<span><span class="co">#&gt; number of iterations= 49 </span></span>
<span><span class="co">#&gt; number of iterations= 73</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">$</span><span class="va">hash.ID</span>,<span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">doubletClassify_between_label</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kable_styling</a></span><span class="op">(</span>bootstrap_options <span class="op">=</span> <span class="st">"bordered"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1
</th>
<th style="text-align:right;">
2
</th>
<th style="text-align:right;">
3
</th>
<th style="text-align:right;">
4
</th>
<th style="text-align:right;">
5
</th>
<th style="text-align:right;">
6
</th>
<th style="text-align:right;">
doublet/multiplet
</th>
<th style="text-align:right;">
negative
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Doublet
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
633
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag1
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
41
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
19
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
166
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
98
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
121
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
37
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
109
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Hashtag6
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
206
</td>
<td style="text-align:right;">
177
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Negative
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
8
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">seurat</span>,<span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">multiseq</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            Doublet Hashtag1 Hashtag2 Hashtag3 Hashtag4 Hashtag5 Hashtag6</span></span>
<span><span class="co">#&gt;   Doublet      445       13        5       32       22       41       70</span></span>
<span><span class="co">#&gt;   Hashtag1       4       99        0        0        0        6        0</span></span>
<span><span class="co">#&gt;   Hashtag2       1        0       20        0        0        3        0</span></span>
<span><span class="co">#&gt;   Hashtag3       9        0        0      204        0        2        0</span></span>
<span><span class="co">#&gt;   Hashtag4       6        0        0        0      135        0        0</span></span>
<span><span class="co">#&gt;   Hashtag5       0        0        0        0        0      177        0</span></span>
<span><span class="co">#&gt;   Hashtag6      12        0        0        0        0        0      353</span></span>
<span><span class="co">#&gt;   Negative       0        0        0        0        0      129        0</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            Negative</span></span>
<span><span class="co">#&gt;   Doublet         5</span></span>
<span><span class="co">#&gt;   Hashtag1       12</span></span>
<span><span class="co">#&gt;   Hashtag2        5</span></span>
<span><span class="co">#&gt;   Hashtag3       49</span></span>
<span><span class="co">#&gt;   Hashtag4       17</span></span>
<span><span class="co">#&gt;   Hashtag5        0</span></span>
<span><span class="co">#&gt;   Hashtag6       18</span></span>
<span><span class="co">#&gt;   Negative      106</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">multiseq</span>,<span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">doubletClassify_between_label</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;              1   2   3   4   5   6 doublet/multiplet negative</span></span>
<span><span class="co">#&gt;   Doublet    0   0   0   0   0   0               477        0</span></span>
<span><span class="co">#&gt;   Hashtag1  68   0   0   0   0   0                44        0</span></span>
<span><span class="co">#&gt;   Hashtag2   0  15   0   0   0   0                10        0</span></span>
<span><span class="co">#&gt;   Hashtag3   0   0 125   0   0   0               111        0</span></span>
<span><span class="co">#&gt;   Hashtag4   0   0   0 109   0   0                48        0</span></span>
<span><span class="co">#&gt;   Hashtag5   0   0   0   0 202   0               156        0</span></span>
<span><span class="co">#&gt;   Hashtag6   0   0   0   0   0 193               230        0</span></span>
<span><span class="co">#&gt;   Negative  13   5  61  13  53  16                43        8</span></span></code></pre></div>
<p>For the remainder of this vignette we will outline our method of
checking whether or not these cells have been called correctly and how
to assign them to their appropriate group!</p>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>Common variants files, for example from the 1000 Genomes Project, can
contain over 7 million SNPs. To reduce computational cost and cell-type
effects, we subset our SNPs list to those located within genes expressed
across most cells in our data.</p>
<p>We first find the most commonly expressed genes in our RNA data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/common_genes.html" class="external-link">common_genes</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">multiplexed_scrnaseq_sce</span>, n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] "TPT1"   "RPL13"  "RPL28"  "TMSB4X" "RPS27"  "EEF1A1" "RPL41"  "B2M"   </span></span>
<span><span class="co">#&gt;  [9] "RPLP1"  "RPL32"</span></span></code></pre></div>
<p>We have a sample .vcf preloaded, but you can load your .vcf file in
using ‘readVcf()’ from <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="external-link">VariantAnnotation</a>.</p>
<p>We will subset our .vcf file to SNPs seen in commonly expressed genes
from our dataset. Notice that the genome for the vcf and EnsDb object
must be compatible!</p>
<p>The returned vcf can be written to file and used with VarTrix later
on.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ensdb</span> <span class="op">&lt;-</span> <span class="fu">EnsDb.Hsapiens.v86</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/EnsDb.Hsapiens.v86/man/package.html" class="external-link">EnsDb.Hsapiens.v86</a></span></span>
<span></span>
<span><span class="fu">genome</span><span class="op">(</span><span class="va">commonvariants_1kgenomes_subset</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">==</span> <span class="fu">genome</span><span class="op">(</span><span class="va">ensdb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt;    1 </span></span>
<span><span class="co">#&gt; TRUE</span></span>
<span></span>
<span><span class="va">new_vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/subset_vcf.html" class="external-link">subset_vcf</a></span><span class="op">(</span><span class="va">commonvariants_1kgenomes_subset</span>, top_genes <span class="op">=</span> <span class="va">top_genes</span>, <span class="va">ensdb</span><span class="op">)</span></span>
<span><span class="va">commonvariants_1kgenomes_subset</span></span>
<span><span class="co">#&gt; class: CollapsedVCF </span></span>
<span><span class="co">#&gt; dim: 2609 0 </span></span>
<span><span class="co">#&gt; rowRanges(vcf):</span></span>
<span><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span><span class="co">#&gt; info(vcf):</span></span>
<span><span class="co">#&gt;   DataFrame with 1 column: AF</span></span>
<span><span class="co">#&gt; info(header(vcf)):</span></span>
<span><span class="co">#&gt;       Number Type  Description                                  </span></span>
<span><span class="co">#&gt;    AF A      Float Estimated allele frequency in the range (0,1)</span></span>
<span><span class="co">#&gt; geno(vcf):</span></span>
<span><span class="co">#&gt;   List of length 0:</span></span>
<span><span class="va">new_vcf</span></span>
<span><span class="co">#&gt; class: CollapsedVCF </span></span>
<span><span class="co">#&gt; dim: 2399 0 </span></span>
<span><span class="co">#&gt; rowRanges(vcf):</span></span>
<span><span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span><span class="co">#&gt; info(vcf):</span></span>
<span><span class="co">#&gt;   DataFrame with 1 column: AF</span></span>
<span><span class="co">#&gt; info(header(vcf)):</span></span>
<span><span class="co">#&gt;       Number Type  Description                                  </span></span>
<span><span class="co">#&gt;    AF A      Float Estimated allele frequency in the range (0,1)</span></span>
<span><span class="co">#&gt; geno(vcf):</span></span>
<span><span class="co">#&gt;   List of length 0:</span></span></code></pre></div>
<p>The subsetted .vcf can be written to disk using ‘writeVcf()’, again
from <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="external-link">VariantAnnotation</a>
package.</p>
<p>Next, we wish to identify cells which we can confidently call to a
particular group. There are a number of ways this can be achieved,
including probabilistic modelling of the HTO counts, manually setting
non-conservative thresholds or using consensus calls. The user may wish
to experiment with different approaches. Here we have used <a href="https://bioconductor.org/packages/release/bioc/html/demuxmix.html" class="external-link">demuxmix</a>,
a probabilistic model which we have set with a high acceptance threshold
to identify high confidence cell calls to use as training data (cells
which we can confidently call as a particular singlet group).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiplexed_scrnaseq_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/high_conf_calls.html" class="external-link">high_conf_calls</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">train</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;   955  1045</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">predict</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 2000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Hashtag1  Hashtag2  Hashtag3  Hashtag4  Hashtag5  Hashtag6 multiplet  negative </span></span>
<span><span class="co">#&gt;        62        15       226       102       348       292       335        12 </span></span>
<span><span class="co">#&gt; uncertain </span></span>
<span><span class="co">#&gt;       608</span></span></code></pre></div>
<p>So, for this particular dataset, we can confidently call 1,045 cells
as being from a particular singlet group. 608 cells cannot be called to
a group with high confidence.</p>
</div>
<div class="section level2">
<h2 id="variant-calling-vartrix">Variant Calling (VarTrix)<a class="anchor" aria-label="anchor" href="#variant-calling-vartrix"></a>
</h2>
<p>Variant calling is not done within the package. Instead, we refer the
reader to <a href="https://github.com/10XGenomics/vartrix" class="external-link">VarTrix</a>,
where they can use the subsetted .vcf file along with their .bam,
barcodes.tsv and reference genome to call SNPs in each cell.</p>
<p>A sample VarTrix command looks like the following:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="ex">./vartrix</span> <span class="at">-v</span> <span class="op">&lt;</span>path_to_input_vcf<span class="op">&gt;</span> -b <span class="op">&lt;</span>path_to_cellranger_bam<span class="op">&gt;</span> -f <span class="op">&lt;</span>path_to_fasta_file<span class="op">&gt;</span> -c <span class="op">&lt;</span>path_to_cell_barcodes_file<span class="op">&gt;</span> -o <span class="op">&lt;</span>path_for_output_matrix<span class="op">&gt;</span></span></code></pre></div>
<p>Using the output matrix from Vartrix and the high confidence
classifications from the HTO algorithm, we can reassign cells using
k-nearest neighbours.</p>
</div>
<div class="section level2">
<h2 id="cell-reassignment-visualisation-and-evaluation">Cell Reassignment, Visualisation and Evaluation<a class="anchor" aria-label="anchor" href="#cell-reassignment-visualisation-and-evaluation"></a>
</h2>
<p>To keep things tidy, we will add the SNP data to our
SingleCellExperiment object as an altExp. We recode the SNP matrix as
follows: 0=no read, 1=SNP present, -1=SNP absent. This function also
filters out SNPs which are observed at a low frequency in the data, and
the frequency threshold can be set manually.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">vartrix_consensus_snps</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2542 2000</span></span>
<span></span>
<span><span class="va">multiplexed_scrnaseq_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/add_snps.html" class="external-link">add_snps</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>, <span class="va">vartrix_consensus_snps</span>, thresh <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>, <span class="st">"SNP"</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 85 2000 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(85): Snp Snp ... Snp Snp</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(2000): AAACCTGAGATCTGCT-1 AAACCTGAGCGTCAAG-1 ...</span></span>
<span><span class="co">#&gt;   ACTTTCAGTAAGTTCC-1 ACTTTCAGTAGGCATG-1</span></span>
<span><span class="co">#&gt; colData names(0):</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>Before we reassign any cells, we will first use the SNPs data to
inspect the results from stand-alone algorithms. Splitting the SNP data
by Seurat HTODemux classification, we initially see a large number of
‘negative’ cells which appear of good quality (high proportion of reads)
which may be assignable to another group. This is consistent with the
library size plot we visualised earlier.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>, <span class="st">"SNP"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    column_split <span class="op">=</span> <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">seurat</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_column_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    column_title_rot <span class="op">=</span> <span class="op">-</span><span class="fl">45</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>    show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"SNP profiles split by Seurat HTODemux call"</span>,</span>
<span>    padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">15</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-18-1.png" width="720"></p>
<p>We will use our knn method to reassign cells based on their SNP
profiles. The training data is the high confidence cells</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">multiplexed_scrnaseq_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/reassign.html" class="external-link">reassign</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>,</span>
<span>    k <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    d <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    train_cells <span class="op">=</span> <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">train</span>,</span>
<span>    predict_cells <span class="op">=</span> <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">predict</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">knn</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Doublet Hashtag1 Hashtag2 Hashtag3 Hashtag4 Hashtag5 Hashtag6 </span></span>
<span><span class="co">#&gt;      337      119       30      349      174      464      527</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>, <span class="st">"SNP"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    column_split <span class="op">=</span> <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">knn</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_column_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    column_names_rot <span class="op">=</span> <span class="fl">45</span>,</span>
<span>    column_title_rot <span class="op">=</span> <span class="op">-</span><span class="fl">45</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>    show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"SNP profiles split by updated knn classification"</span>,</span>
<span>    padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">15</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-20-1.png" width="720"></p>
<p>Focusing in on the knn Hashtag5 group, we see that a lot of the
Negative cells have now been correctly reclassed to this group, as well
as a small number of cells from other groups.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">multiplexed_scrnaseq_sce</span>, <span class="st">"SNP"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">knn</span> <span class="op">==</span> <span class="st">"Hashtag5"</span><span class="op">]</span>,</span>
<span>    column_split <span class="op">=</span> <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">seurat</span><span class="op">[</span><span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">knn</span> <span class="op">==</span> <span class="st">"Hashtag5"</span><span class="op">]</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_column_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    column_names_rot <span class="op">=</span> <span class="fl">45</span>,</span>
<span>    column_title_rot <span class="op">=</span> <span class="op">-</span><span class="fl">45</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>    show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"knn Hashtag5 group split by Seurat HTODemux classification"</span>,</span>
<span>    padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">15</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-21-1.png" width="720"></p>
<div class="section level3">
<h3 id="performance">Performance<a class="anchor" aria-label="anchor" href="#performance"></a>
</h3>
<p>Next we will run some basic performance checks. We subset our
SingleCellExperiment object to only retain cells which we could
confidently call as being from a singlet group, then split this into a
training and test dataset.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_test</span> <span class="op">&lt;-</span> <span class="va">multiplexed_scrnaseq_sce</span><span class="op">[</span>, <span class="va">multiplexed_scrnaseq_sce</span><span class="op">$</span><span class="va">train</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="va">sce_test</span><span class="op">$</span><span class="va">knn</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span><span class="va">sce_test</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 259 1045 </span></span>
<span><span class="co">#&gt; metadata(2): doubletClassify_between_threshold</span></span>
<span><span class="co">#&gt;   doubletClassify_between_resultsMat</span></span>
<span><span class="co">#&gt; assays(2): counts logcounts</span></span>
<span><span class="co">#&gt; rownames(259): RPL22 CDC42 ... MT-ND5 MT-CYB</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1045): AAACCTGAGCGTCAAG-1 AAACCTGAGGCGTACA-1 ...</span></span>
<span><span class="co">#&gt;   ACTTTCAGTAAGAGAG-1 ACTTTCAGTAAGTTCC-1</span></span>
<span><span class="co">#&gt; colData names(14): orig.ident nCount_RNA ... predict labels</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: RNA</span></span>
<span><span class="co">#&gt; altExpNames(2): HTO SNP</span></span>
<span></span>
<span><span class="va">sce_test</span><span class="op">$</span><span class="va">train2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">train</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce_test</span><span class="op">$</span><span class="va">train2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span></span>
<span><span class="va">sce_test</span><span class="op">$</span><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">sce_test</span><span class="op">$</span><span class="va">train2</span> <span class="op">==</span> <span class="cn">FALSE</span></span></code></pre></div>
<p>Comparing the predicted labels in the test dataset with the hidden
high confidence labels, we see excellent agreement.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/reassign.html" class="external-link">reassign</a></span><span class="op">(</span><span class="va">sce_test</span>, k <span class="op">=</span> <span class="fl">3</span>, train_cells <span class="op">=</span> <span class="va">sce_test</span><span class="op">$</span><span class="va">train2</span>, predict_cells <span class="op">=</span> <span class="va">sce_test</span><span class="op">$</span><span class="va">test</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span><span class="op">[</span><span class="va">sce_test</span><span class="op">$</span><span class="va">test</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">]</span>, <span class="va">sce_test</span><span class="op">$</span><span class="va">knn</span><span class="op">[</span><span class="va">sce_test</span><span class="op">$</span><span class="va">test</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            Doublet Hashtag1 Hashtag2 Hashtag3 Hashtag4 Hashtag5 Hashtag6</span></span>
<span><span class="co">#&gt;   Hashtag1       1       35        0        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag2       1        0       10        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag3       0        0        0      122        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag4       0        0        0        0       46        0        0</span></span>
<span><span class="co">#&gt;   Hashtag5       1        0        0        0        0      193        1</span></span>
<span><span class="co">#&gt;   Hashtag6       0        0        0        0        0        0      135</span></span></code></pre></div>
<p>We can also show that that the model can correct misclassified cells
when predicted back on the training data. We create a new vector
“labels2” which is initially identical to the labels used in training
previously.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_test</span><span class="op">$</span><span class="va">knn</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span>, <span class="va">sce_test</span><span class="op">$</span><span class="va">labels2</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            Hashtag1 Hashtag2 Hashtag3 Hashtag4 Hashtag5 Hashtag6</span></span>
<span><span class="co">#&gt;   Hashtag1       62        0        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag2        0       15        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag3        0        0      226        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag4        0        0        0      102        0        0</span></span>
<span><span class="co">#&gt;   Hashtag5        0        0        0        0      348        0</span></span>
<span><span class="co">#&gt;   Hashtag6        0        0        0        0        0      292</span></span></code></pre></div>
<p>We then randomly reassign 25 cells from Hashtag5 to Hashtag2 in the
training data. Predicting the model back on itself we see that in the
new predicted labels, the 25 altered have been correctly reclassified
back to Hashtag5</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels2</span> <span class="op">==</span> <span class="st">"Hashtag5"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Hashtag2"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span>, <span class="va">sce_test</span><span class="op">$</span><span class="va">labels2</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            Hashtag1 Hashtag2 Hashtag3 Hashtag4 Hashtag5 Hashtag6</span></span>
<span><span class="co">#&gt;   Hashtag1       62        0        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag2        0       15        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag3        0        0      226        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag4        0        0        0      102        0        0</span></span>
<span><span class="co">#&gt;   Hashtag5        0       25        0        0      323        0</span></span>
<span><span class="co">#&gt;   Hashtag6        0        0        0        0        0      292</span></span>
<span></span>
<span><span class="va">sce_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demuxSNP/man/reassign.html" class="external-link">reassign</a></span><span class="op">(</span><span class="va">sce_test</span>,</span>
<span>    train_cells <span class="op">=</span> <span class="va">sce_test</span><span class="op">$</span><span class="va">train</span>,</span>
<span>    predict_cells <span class="op">=</span> <span class="va">sce_test</span><span class="op">$</span><span class="va">train</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span>, <span class="va">sce_test</span><span class="op">$</span><span class="va">knn</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            Doublet Hashtag1 Hashtag2 Hashtag3 Hashtag4 Hashtag5 Hashtag6</span></span>
<span><span class="co">#&gt;   Hashtag1       0       62        0        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag2       0        0       15        0        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag3       0        0        0      226        0        0        0</span></span>
<span><span class="co">#&gt;   Hashtag4       0        0        0        0      102        0        0</span></span>
<span><span class="co">#&gt;   Hashtag5       1        0        0        0        0      346        1</span></span>
<span><span class="co">#&gt;   Hashtag6       0        0        0        0        0        0      292</span></span></code></pre></div>
<p>For the knn Hashtag6 group, we see mostly good agreement with the
original demuxmix labels. However, it appears one cell now being called
a Hashtag6 was originally called Hashtag5. Again, we can investigate
this by visualising the corresponding SNP profile. In doing so, we can
see that this most likely is a true Hashtag6.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce_test</span>, <span class="st">"SNP"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="va">sce_test</span><span class="op">$</span><span class="va">knn</span> <span class="op">==</span> <span class="st">"Hashtag6"</span><span class="op">]</span>,</span>
<span>    column_split <span class="op">=</span> <span class="va">sce_test</span><span class="op">$</span><span class="va">labels</span><span class="op">[</span><span class="va">sce_test</span><span class="op">$</span><span class="va">knn</span> <span class="op">==</span> <span class="st">"Hashtag6"</span><span class="op">]</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_column_slices <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    column_names_rot <span class="op">=</span> <span class="fl">45</span>,</span>
<span>    column_title_rot <span class="op">=</span> <span class="op">-</span><span class="fl">45</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"SNPs"</span>,</span>
<span>    show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"knn Hashtag6 group split by demuxmix classification"</span>,</span>
<span>    padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">15</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="c-demuxSNP-package-demo_files/figure-html/unnamed-chunk-26-1.png" width="720"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] CiteFuse_1.12.0             kableExtra_1.3.4           </span></span>
<span><span class="co">#&gt;  [3] EnsDb.Hsapiens.v86_2.99.0   dittoSeq_1.12.0            </span></span>
<span><span class="co">#&gt;  [5] ggpubr_0.6.0                ggplot2_3.4.2              </span></span>
<span><span class="co">#&gt;  [7] SeuratObject_4.1.3          Seurat_4.3.0.1             </span></span>
<span><span class="co">#&gt;  [9] viridisLite_0.4.2           ComplexHeatmap_2.16.0      </span></span>
<span><span class="co">#&gt; [11] demuxSNP_0.99.10            ensembldb_2.24.0           </span></span>
<span><span class="co">#&gt; [13] AnnotationFilter_1.24.0     GenomicFeatures_1.52.1     </span></span>
<span><span class="co">#&gt; [15] AnnotationDbi_1.62.2        VariantAnnotation_1.46.0   </span></span>
<span><span class="co">#&gt; [17] Rsamtools_2.16.0            Biostrings_2.68.1          </span></span>
<span><span class="co">#&gt; [19] XVector_0.40.0              SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt; [21] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt; [23] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">#&gt; [25] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">#&gt; [27] BiocGenerics_0.46.0         MatrixGenerics_1.12.2      </span></span>
<span><span class="co">#&gt; [29] matrixStats_1.0.0           BiocStyle_2.28.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] progress_1.2.2            goftest_1.2-3            </span></span>
<span><span class="co">#&gt;   [3] vctrs_0.6.3               spatstat.random_3.1-5    </span></span>
<span><span class="co">#&gt;   [5] digest_0.6.33             png_0.1-8                </span></span>
<span><span class="co">#&gt;   [7] shape_1.4.6               ggrepel_0.9.3            </span></span>
<span><span class="co">#&gt;   [9] deldir_1.0-9              parallelly_1.36.0        </span></span>
<span><span class="co">#&gt;  [11] combinat_0.0-8            magick_2.7.4             </span></span>
<span><span class="co">#&gt;  [13] MASS_7.3-60               pkgdown_2.0.7            </span></span>
<span><span class="co">#&gt;  [15] reshape2_1.4.4            httpuv_1.6.11            </span></span>
<span><span class="co">#&gt;  [17] foreach_1.5.2             withr_2.5.0              </span></span>
<span><span class="co">#&gt;  [19] xfun_0.39                 ellipsis_0.3.2           </span></span>
<span><span class="co">#&gt;  [21] survival_3.5-5            memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [23] ggsci_3.0.0               mixtools_2.0.0           </span></span>
<span><span class="co">#&gt;  [25] systemfonts_1.0.4         ragg_1.2.5               </span></span>
<span><span class="co">#&gt;  [27] zoo_1.8-12                GlobalOptions_0.1.2      </span></span>
<span><span class="co">#&gt;  [29] pbapply_1.7-2             DEoptimR_1.1-0           </span></span>
<span><span class="co">#&gt;  [31] prettyunits_1.1.1         KEGGREST_1.40.0          </span></span>
<span><span class="co">#&gt;  [33] promises_1.2.0.1          httr_1.4.6               </span></span>
<span><span class="co">#&gt;  [35] rstatix_0.7.2             restfulr_0.0.15          </span></span>
<span><span class="co">#&gt;  [37] globals_0.16.2            fitdistrplus_1.1-11      </span></span>
<span><span class="co">#&gt;  [39] rhdf5filters_1.12.1       demuxmix_1.2.0           </span></span>
<span><span class="co">#&gt;  [41] rhdf5_2.44.0              rstudioapi_0.15.0        </span></span>
<span><span class="co">#&gt;  [43] miniUI_0.1.1.1            generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [45] curl_5.0.1                zlibbioc_1.46.0          </span></span>
<span><span class="co">#&gt;  [47] ScaledMatrix_1.8.1        ggraph_2.1.0             </span></span>
<span><span class="co">#&gt;  [49] polyclip_1.10-4           randomForest_4.7-1.1     </span></span>
<span><span class="co">#&gt;  [51] GenomeInfoDbData_1.2.10   xtable_1.8-4             </span></span>
<span><span class="co">#&gt;  [53] stringr_1.5.0             desc_1.4.2               </span></span>
<span><span class="co">#&gt;  [55] doParallel_1.0.17         evaluate_0.21            </span></span>
<span><span class="co">#&gt;  [57] S4Arrays_1.0.4            BiocFileCache_2.8.0      </span></span>
<span><span class="co">#&gt;  [59] hms_1.1.3                 bookdown_0.34            </span></span>
<span><span class="co">#&gt;  [61] irlba_2.3.5.1             colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [63] filelock_1.0.2            ROCR_1.0-11              </span></span>
<span><span class="co">#&gt;  [65] reticulate_1.30           spatstat.data_3.0-1      </span></span>
<span><span class="co">#&gt;  [67] magrittr_2.0.3            lmtest_0.9-40            </span></span>
<span><span class="co">#&gt;  [69] later_1.3.1               viridis_0.6.3            </span></span>
<span><span class="co">#&gt;  [71] lattice_0.21-8            spatstat.geom_3.2-2      </span></span>
<span><span class="co">#&gt;  [73] future.apply_1.11.0       robustbase_0.99-0        </span></span>
<span><span class="co">#&gt;  [75] scuttle_1.10.1            scattermore_1.2          </span></span>
<span><span class="co">#&gt;  [77] XML_3.99-0.14             cowplot_1.1.1            </span></span>
<span><span class="co">#&gt;  [79] RcppAnnoy_0.0.21          class_7.3-22             </span></span>
<span><span class="co">#&gt;  [81] pillar_1.9.0              nlme_3.1-162             </span></span>
<span><span class="co">#&gt;  [83] iterators_1.0.14          compiler_4.3.1           </span></span>
<span><span class="co">#&gt;  [85] beachmat_2.16.0           stringi_1.7.12           </span></span>
<span><span class="co">#&gt;  [87] tensor_1.5                GenomicAlignments_1.36.0 </span></span>
<span><span class="co">#&gt;  [89] plyr_1.8.8                crayon_1.5.2             </span></span>
<span><span class="co">#&gt;  [91] abind_1.4-5               BiocIO_1.10.0            </span></span>
<span><span class="co">#&gt;  [93] locfit_1.5-9.8            sp_2.0-0                 </span></span>
<span><span class="co">#&gt;  [95] graphlayouts_1.0.0        bit_4.0.5                </span></span>
<span><span class="co">#&gt;  [97] dplyr_1.1.2               codetools_0.2-19         </span></span>
<span><span class="co">#&gt;  [99] textshaping_0.3.6         BiocSingular_1.16.0      </span></span>
<span><span class="co">#&gt; [101] bslib_0.5.0               GetoptLong_1.0.5         </span></span>
<span><span class="co">#&gt; [103] plotly_4.10.2             mime_0.12                </span></span>
<span><span class="co">#&gt; [105] splines_4.3.1             circlize_0.4.15          </span></span>
<span><span class="co">#&gt; [107] Rcpp_1.0.11               dbplyr_2.3.3             </span></span>
<span><span class="co">#&gt; [109] sparseMatrixStats_1.12.2  knitr_1.43               </span></span>
<span><span class="co">#&gt; [111] blob_1.2.4                utf8_1.2.3               </span></span>
<span><span class="co">#&gt; [113] clue_0.3-64               fs_1.6.2                 </span></span>
<span><span class="co">#&gt; [115] listenv_0.9.0             DelayedMatrixStats_1.22.1</span></span>
<span><span class="co">#&gt; [117] ggsignif_0.6.4            tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [119] Matrix_1.6-0              statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [121] svglite_2.1.1             tweenr_2.0.2             </span></span>
<span><span class="co">#&gt; [123] pkgconfig_2.0.3           pheatmap_1.0.12          </span></span>
<span><span class="co">#&gt; [125] tools_4.3.1               cachem_1.0.8             </span></span>
<span><span class="co">#&gt; [127] RSQLite_2.3.1             rvest_1.0.3              </span></span>
<span><span class="co">#&gt; [129] DBI_1.1.3                 fastmap_1.1.1            </span></span>
<span><span class="co">#&gt; [131] rmarkdown_2.23            scales_1.2.1             </span></span>
<span><span class="co">#&gt; [133] ica_1.0-3                 broom_1.0.5              </span></span>
<span><span class="co">#&gt; [135] sass_0.4.7                patchwork_1.1.2          </span></span>
<span><span class="co">#&gt; [137] BiocManager_1.30.21       carData_3.0-5            </span></span>
<span><span class="co">#&gt; [139] RANN_2.6.1                farver_2.1.1             </span></span>
<span><span class="co">#&gt; [141] tidygraph_1.2.3           yaml_2.3.7               </span></span>
<span><span class="co">#&gt; [143] bayesm_3.1-5              rtracklayer_1.60.0       </span></span>
<span><span class="co">#&gt; [145] cli_3.6.1                 purrr_1.0.1              </span></span>
<span><span class="co">#&gt; [147] webshot_0.5.5             leiden_0.4.3             </span></span>
<span><span class="co">#&gt; [149] lifecycle_1.0.3           dbscan_1.1-11            </span></span>
<span><span class="co">#&gt; [151] uwot_0.1.16               bluster_1.10.0           </span></span>
<span><span class="co">#&gt; [153] kernlab_0.9-32            backports_1.4.1          </span></span>
<span><span class="co">#&gt; [155] BiocParallel_1.34.2       gtable_0.3.3             </span></span>
<span><span class="co">#&gt; [157] rjson_0.2.21              ggridges_0.5.4           </span></span>
<span><span class="co">#&gt; [159] progressr_0.13.0          parallel_4.3.1           </span></span>
<span><span class="co">#&gt; [161] limma_3.56.2              jsonlite_1.8.7           </span></span>
<span><span class="co">#&gt; [163] edgeR_3.42.4              bitops_1.0-7             </span></span>
<span><span class="co">#&gt; [165] bit64_4.0.5               Rtsne_0.16               </span></span>
<span><span class="co">#&gt; [167] spatstat.utils_3.0-3      BiocNeighbors_1.18.0     </span></span>
<span><span class="co">#&gt; [169] metapod_1.8.0             jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt; [171] highr_0.10                dqrng_0.3.0              </span></span>
<span><span class="co">#&gt; [173] segmented_1.6-4           lazyeval_0.2.2           </span></span>
<span><span class="co">#&gt; [175] shiny_1.7.4.1             htmltools_0.5.5          </span></span>
<span><span class="co">#&gt; [177] sctransform_0.3.5         rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt; [179] glue_1.6.2                RCurl_1.98-1.12          </span></span>
<span><span class="co">#&gt; [181] rprojroot_2.0.3           scran_1.28.1             </span></span>
<span><span class="co">#&gt; [183] BSgenome_1.68.0           gridExtra_2.3            </span></span>
<span><span class="co">#&gt; [185] igraph_1.5.0              R6_2.5.1                 </span></span>
<span><span class="co">#&gt; [187] tidyr_1.3.0               labeling_0.4.2           </span></span>
<span><span class="co">#&gt; [189] cluster_2.1.4             Rhdf5lib_1.22.0          </span></span>
<span><span class="co">#&gt; [191] compositions_2.0-6        DelayedArray_0.26.6      </span></span>
<span><span class="co">#&gt; [193] tidyselect_1.2.0          ProtGenerics_1.32.0      </span></span>
<span><span class="co">#&gt; [195] tensorA_0.36.2            ggforce_0.4.1            </span></span>
<span><span class="co">#&gt; [197] xml2_1.3.5                car_3.1-2                </span></span>
<span><span class="co">#&gt; [199] future_1.33.0             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt; [201] munsell_0.5.0             KernSmooth_2.23-21       </span></span>
<span><span class="co">#&gt; [203] data.table_1.14.8         htmlwidgets_1.6.2        </span></span>
<span><span class="co">#&gt; [205] RColorBrewer_1.1-3        biomaRt_2.56.1           </span></span>
<span><span class="co">#&gt; [207] rlang_1.1.1               spatstat.sparse_3.0-2    </span></span>
<span><span class="co">#&gt; [209] spatstat.explore_3.2-1    Cairo_1.6-0              </span></span>
<span><span class="co">#&gt; [211] fansi_1.0.4</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Lynch.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
